src/Build/BackEnd/Components/RequestBuilder/TaskRoutingDecision.cs

```diff
@@ -0,0 +1,116 @@
+        /// - Tasks without these indicators must run in a sidecar TaskHost for isolation (returns true)
+        /// </remarks>
+        public static bool NeedsTaskHostInMultiThreadedMode(Type taskType)
+        {
+            ErrorUtilities.VerifyThrowArgumentNull(taskType, nameof(taskType));
+
+            // Tasks without thread-safety indicators need isolation in a TaskHost sidecar
+            return !IsMultiThreadableTask(taskType);
+        }
+
+        /// <summary>
+        /// Checks if a task is multi-threadable via IMultiThreadableTask interface OR
+        /// MSBuildMultiThreadableTaskAttribute attribute.
+        /// Results are cached to avoid repeated reflection calls.
+        /// </summary>
+        /// <param name="taskType">The task type to check.</param>
+        /// <returns>True if the task has thread-safety indicators; false otherwise.</returns>
+        private static bool IsMultiThreadableTask(Type taskType)
+        {
+            return s_multiThreadableTaskCache.GetOrAdd(
+                taskType,
+                static t =>
+                    ImplementsIMultiThreadableTask(t) ||
+                    HasMultiThreadableTaskAttribute(t));
+        }
+
+        /// <summary>
+        /// Checks if a task type implements the IMultiThreadableTask interface.
+        /// </summary>
+        /// <param name="taskType">The task type to check.</param>
+        /// <returns>True if the task implements IMultiThreadableTask; false otherwise.</returns>
+        private static bool ImplementsIMultiThreadableTask(Type taskType)
+        {
+            return typeof(IMultiThreadableTask).IsAssignableFrom(taskType);
+        }
+
+        /// <summary>
+        /// Checks if a task type is marked with MSBuildMultiThreadableTaskAttribute.
+        /// Detection is based on namespace and name only, ignoring the defining assembly,
+        /// which allows customers to define the attribute in their own assemblies.
+        /// </summary>
+        /// <param name="taskType">The task type to check.</param>
+        /// <returns>True if the task has the attribute; false otherwise.</returns>
+        private static bool HasMultiThreadableTaskAttribute(Type taskType)
+        {
+            const string attributeFullName = "Microsoft.Build.Framework.MSBuildMultiThreadableTaskAttribute";
+
+            // Check for the attribute by full name, not by type identity
+            // This allows custom-defined attributes from different assemblies
+            foreach (object attr in taskType.GetCustomAttributes(inherit: false))
+            {
+                if (attr.GetType().FullName == attributeFullName)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Clears the thread-safety capability cache. Used primarily for testing.
+        /// </summary>
+        internal static void ClearCache()
```

Comment:
You will need to call it when build finishes for cache to work correctly with MSBuild server.